test/test_util/tpcc/workload_cached.cpp

```diff
@@ -0,0 +1,239 @@
+  db_main_ = db_main;
+  txn_manager_ = db_main_->GetTransactionLayer()->GetTransactionManager();
+  block_store_ = db_main_->GetStorageLayer()->GetBlockStore();
+  catalog_ = db_main_->GetCatalogLayer()->GetCatalog();
+  txn_manager_ = db_main_->GetTransactionLayer()->GetTransactionManager();
+
+  // build the TPCC database using HashMaps where possible
+  Builder tpcc_builder{block_store_, catalog_, txn_manager_};
+  tpcc_db_ = tpcc_builder.Build(storage::index::IndexType::HASHMAP);
+  db_oid_ = tpcc_db_->db_oid_;
+
+  GenerateTPCCTables(num_threads);
+
+  // initialize and compile the queries
+  InitializeSQLs();
+  LoadTPCCQueries(txn_names);
+}
+
+void WorkloadCached::GenerateTPCCTables(int8_t num_threads) {
+  // populate the tables and indexes
+  // prepare the thread pool and workers
+  common::WorkerPool thread_pool{static_cast<uint32_t>(num_threads), {}};
+  std::vector<Worker> workers;
+  workers.reserve(num_threads);
+  workers.clear();
+  for (int8_t i = 0; i < num_threads; i++) {
+    workers.emplace_back(tpcc_db_);
+  }
+  thread_pool.Startup();
+  Loader::PopulateDatabase(txn_manager_, tpcc_db_, &workers, &thread_pool);
+}
+
+void WorkloadCached::LoadTPCCQueries(const std::vector<std::string> &txn_names) {
+  uint64_t optimizer_timeout =
+      static_cast<uint64_t>(db_main_->GetSettingsManager()->GetInt(settings::Param::task_execution_timeout));
+
+  for (auto &txn_name : txn_names) {
+    // read queries from files
+    auto curr = queries_.emplace(txn_name, std::vector<execution::ExecutableQuery>{});
+
+    for (auto &query : sqls_.find(txn_name)->second) {
+      const auto parse_result = parser::PostgresParser::BuildParseTree(query);
+      transaction::TransactionContext *txn = txn_manager_->BeginTransaction();
+
+      auto accessor = catalog_->GetAccessor(common::ManagedPointer(txn), db_oid_);
+      binder::BindNodeVisitor visitor(common::ManagedPointer(accessor.get()), db_oid_);
+      visitor.BindNameToNode(common::ManagedPointer<parser::ParseResult>(parse_result));
+
+      // generate plan node
+      std::unique_ptr<planner::AbstractPlanNode> plan_node = trafficcop::TrafficCopUtil::Optimize(
+          common::ManagedPointer(txn), common::ManagedPointer(accessor), common::ManagedPointer(parse_result), db_oid_,
+          db_main_->GetStatsStorage(), std::make_unique<optimizer::TrivialCostModel>(), optimizer_timeout);
+
+      auto exec_ctx = std::make_unique<execution::exec::ExecutionContext>(db_oid_, common::ManagedPointer(txn), nullptr,
+                                                                          nullptr, common::ManagedPointer(accessor));
+
+      // generate executable query and emplace it into the vector; break down here
+      execution::ExecutableQuery executable{common::ManagedPointer(plan_node), common::ManagedPointer(exec_ctx)};
+      curr.first->second.emplace_back(std::move(executable));
+
+      txn_manager_->Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
```

Comment:
Does running this cause any heap-use-after-freed errors? My understanding is that `txn_manager_->Commit` could end up causing certain plan components (i.e scan predicate) to be freed by deferred events.

Reply:
I'm not sure about the predicate freeing part. But regardless, what would be used after being freed that you're referring to? The plan is only used to compile the TPL modules in the `ExecutableQuery`. We do not save it anyway.

Reply:
Ah ok. Thanks for the clarification.